pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go

```diff
@@ -0,0 +1,100 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func TestAdmit(t *testing.T) {
+	nsName := "test"
+
+	nsExistMutator := func() Mutator {
+		ctrl := gomock.NewController(t)
+		mockNamespaceController := fake.NewMockNonNamespacedControllerInterface[*v1.Namespace, *v1.NamespaceList](ctrl)
+		mockNamespaceController.EXPECT().Create(gomock.Any()).Return(nil, errors.NewAlreadyExists(schema.GroupResource{}, nsName))
+		mockNamespaceController.EXPECT().Get(gomock.Any(), gomock.Any()).Return(&v1.Namespace{}, nil)
+
+		return Mutator{
+			namespaces: mockNamespaceController,
+		}
+	}
+
+	newNsMutator := func() Mutator {
+		ctrl := gomock.NewController(t)
+
+		mockNamespaceController := fake.NewMockNonNamespacedControllerInterface[*v1.Namespace, *v1.NamespaceList](ctrl)
+		mockNamespaceController.EXPECT().Create(gomock.Any()).Return(&v1.Namespace{}, nil)
+		mockNamespaceController.EXPECT().Get(gomock.Any(), gomock.Any()).Times(0)
+		mockRoleBindingController := fake.NewMockControllerInterface[*rbacv1.RoleBinding, *rbacv1.RoleBindingList](ctrl)
+		mockRoleBindingController.EXPECT().Create(gomock.Any())
+		mockClusterRoleBindingController := fake.NewMockNonNamespacedControllerInterface[*rbacv1.ClusterRoleBinding, *rbacv1.ClusterRoleBindingList](ctrl)
+		mockClusterRoleBindingController.EXPECT().Create(gomock.Any())
+		mockClusterRoleController := fake.NewMockNonNamespacedControllerInterface[*rbacv1.ClusterRole, *rbacv1.ClusterRoleList](ctrl)
+		mockClusterRoleController.EXPECT().Create(gomock.Any())
+
+		return Mutator{
+			namespaces:          mockNamespaceController,
+			rolebindings:        mockRoleBindingController,
+			clusterrolebindings: mockClusterRoleBindingController,
+			clusterroles:        mockClusterRoleController,
+		}
+	}
+
+	workspace := v3.FleetWorkspace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nsName,
+		},
+	}
+	raw, _ := json.Marshal(workspace)
+	req := &admission.Request{
+		AdmissionRequest: admissionv1.AdmissionRequest{
+			Object: runtime.RawExtension{Object: &workspace, Raw: raw},
+		},
+	}
+
+	tests := map[string]struct {
+		m                  func() Mutator
+		req                *admission.Request
+		expectAllowed      bool
+		expectResultStatus *metav1.Status
+	}{
+		"reject because namespace already exists": {
+			m:                  nsExistMutator,
+			req:                req,
+			expectAllowed:      false,
+			expectResultStatus: &metav1.Status{Message: "namespace 'test' already exists"},
+		},
+		"new namespace": {
+			m:                  newNsMutator,
+			req:                req,
+			expectAllowed:      true,
+			expectResultStatus: nil,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			m := test.m()
+			admit, _ := m.Admit(test.req)
+			if admit.Allowed != test.expectAllowed {
+				t.Errorf("Response was incorrectly validated, wanted response.Allowed = '%v' got %v: result=%+v", test.expectAllowed, admit.Allowed, admit.Result)
+			}
```

Comment:
We use testify for our condition checking which gives us uniformed error messages between our unit test.
```suggestion
			require.Equalf(test.expectAllowed, admit.Allowed, "Response was incorrectly validated, wanted response.Allowed = '%v' got %v: result=%+v", test.expectAllowed, admit.Allowed, admit.Result)
```

Reply:
done in https://github.com/rancher/webhook/pull/260/commits/bc3715088473c65ec63c87a6da489240e56dbaef